### DISCUSSION TITLE
CORS blocking requests

### DISCUSSION THREAD

**Discussion by leafoliage** [vercel/community#65] [Help] [answered]
I was to fetch data from my own api in next js.
I've added `Access-Control-Allow-Origin: *` to my api response header, however, my requests kept being blocked by CORS.
I got `Access to fetch at 'http://localhost:6001/post/select/hot' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.`  in my console.
Btw, it seems like my api didn't even receive the request.
Anyone knows how this can be solved?

**Comment by styfle** [comment, marked as answer]
Take a look at this guide: https://vercel.com/support/articles/how-to-enable-cors

**Reply by jlew24asu** [reply]
I'm having this same problem.  cors is blocking something, so why do you suggest enabling it?

**Reply by davidklemme** [reply]
If you made it here from google - the answer is in the docs indeed. Depending on your issue, check that you adapt *source* and *destination* according to your use case. 
The standard 
_"// matching all API routes
source: "/api/:path*","_ 
will 'only' apply the rule to all requests coming from your api.

For another example (including multiple header configs) https://vercel.com/docs/cli#project-configuration/headers

**Reply by davidklemme** [reply]
And whatever you do - don't just disable CORS without knowing exactly what you're doing ..

**Reply by ZeeshanAhmadKhalil** [reply]
since i enabled cors in `vercel.json`, its not deploying my builds anymore :/

**Reply by amyegan** [reply]
Sorry you're running into problems @ZeeshanAhmadKhalil. Can you [start a new discussion](https://github.com/vercel/community/discussions/new) with more information about your app? That way we can work together more easily to solve your specific problem.

**Reply by SWS-5007** [reply]
I have followed this guide, [https://vercel.com/support/articles/how-to-enable-cors](https://vercel.com/support/articles/how-to-enable-cors) and changed my `vercel.json` file like below.
But It doesn't solve the problem yet.
Here is my `vercel.json` codebase.

```
{
    "version": 2,
    "functions": {
        "api/index.php": {
            "runtime": "vercel-php@0.3.1"
        }
    },
    "routes": [{
        "src": "/(.*)",
        "dest": "/api/index.php"
    }],
    "headers": [{
        "source": "/api/(.*)",
        "headers": [{
                "key": "Access-Control-Allow-Credentials",
                "value": "true"
            },
            {
                "key": "Access-Control-Allow-Origin",
                "value": "*"
            },
            {
                "key": "Access-Control-Allow-Methods",
                "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT"
            },
            {
                "key": "Access-Control-Allow-Headers",
                "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version"
            }
        ]
    }],
    "env": {
        "APP_NAME": "BusinessCard-Backend",
        "APP_ENV": "production",
        "APP_DEBUG": "false",
        "APP_CONFIG_CACHE": "/tmp/config.php",
        "APP_EVENTS_CACHE": "/tmp/events.php",
        "APP_PACKAGES_CACHE": "/tmp/packages.php",
        "APP_ROUTES_CACHE": "/tmp/routes.php",
        "APP_SERVICES_CACHE": "/tmp/services.php",
        "VIEW_COMPILED_PATH": "/tmp",
        "CACHE_DRIVER": "array",
        "LOG_CHANNEL": "stderr",
        "SESSION_DRIVER": "cookie",
        "VIEW_COMPILED_PATH": "/tmp/views",
        "SSR_TEMP_PATH": "/tmp/ssr",
        "NODE_PATH": "node"
    }
}
```
Please help me!

**Comment by sovoid** [comment]
@styfle I still keep getting the same issue even though I've followed what the documentation says. I currently use firefox developer edition. Here's the code I run:

```js
const btoa = require("btoa");
const path = require("path");
let chromium;
// Check if we are running locally or on Vercel
if (!process.env.AWS_LAMBDA_FUNCTION_VERSION) {
  const puppeteer = require("puppeteer");
  chromium = {
    puppeteer,
  };
} else {
  chromium = require("chrome-aws-lambda");
}
const renderHTML = require("./render-html");

/**
 * Normalizes the theme name by adding a prefix `jsonresume-theme` if not present
 * @param {String} value Name of the theme
 * @param {String} defaultValue Default Theme
 */
const normalizeTheme = (value, defaultValue = "even") => {
  const theme = value || defaultValue;

  // If theme has a relative path provided, immediately return
  if (theme[0] === ".") {
    return theme;
  }

  return theme.match("jsonresume-theme-.*")
    ? theme
    : `jsonresume-theme-${theme}`;
};

/**
 * Returns the theme Package
 * @param {String} theme Normalized theme name
 */
const getThemePackage = (theme) => {
  if (theme[0] === ".") {
    theme = path.join(process.cwd(), theme, "index.js");
  }

  try {
    const themePackage = require(theme);
    return themePackage;
  } catch (err) {
    // Theme not installed
    throw new Error(
      "You have to install this theme relative to the folder to use it e.g. `npm install " +
        theme +
        "`"
    );
  }
};

/**
 * Parses the `resumeJson` to render a pretty HTML
 * @param {Object} resumeJson Resume data
 * @param {String} themeName Name of the theme
 */
const parse = async (resumeJson, themeName) => {
  const html = await renderHTML({
    resume: resumeJson,
    themePath: themeName,
  });
  return html;
};

// See https://vercel.com/knowledge/how-to-enable-cors
const allowCors = (fn) => async (req, res) => {
  res.setHeader("Access-Control-Allow-Credentials", true);
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader(
    "Access-Control-Allow-Methods",
    "GET,OPTIONS,PATCH,DELETE,POST,PUT"
  );
  res.setHeader(
    "Access-Control-Allow-Headers",
    "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version"
  );

  // Handle CORS pre-flight
  if (req.method === "OPTIONS") {
    res.status(200).end();
    return;
  }

  return await fn(req, res);
};

const handler = async (req, res) => {
  try {
    const { resumeData } = req.body;
    const normalizedTheme = normalizeTheme(resumeData.meta.theme, "even");
    const themePackage = getThemePackage(normalizedTheme);
    const html = await parse(resumeData, normalizedTheme);
    const browser = await chromium.puppeteer.launch();
    const page = await browser.newPage();
    await page.emulateMediaType(
      (themePackage.pdfRenderOptions &&
        themePackage.pdfRenderOptions.mediaType) ||
        "screen"
    );
    await page.goto(
      `data:text/html;base64,${btoa(unescape(encodeURIComponent(html)))}`,
      { waitUntil: "networkidle0" }
    );
    if (themePackage.pdfViewport) {
      await page.setViewport(themePackage.pdfViewport);
    }
    const pdf = await page.pdf({
      format: "Letter",
      printBackground: true,
      margin: {
        top: "0.5in",
        bottom: "0.5in",
        left: "0.5in",
        right: "0.5in",
      },
      ...themePackage.pdfRenderOptions,
    });

    await browser.close();

    res.setHeader("Content-Type", "application/pdf");
    res.setHeader("Content-Length", pdf.length);
    res.send(pdf);
  } catch (err) {
    res.status(500);
    res.json({
      error: err,
    });
  }
};

module.exports = allowCors(handler);
```

Let me know if you need more info from my end...any help would be awesome

**Reply by lorstenoplo** [reply]
same here bro, still breaking my head

**Reply by y-nk** [reply]
+1 on this. I've also been trying the `vercel.json` approach in the guide + quick and dirty testing with jsfiddle/codepen. req get stuck at preflight.

**Reply by Muzamil132** [reply]
my head is breaking please help me if you have solved this

**Reply by jlew24asu** [reply]
install the cors browser plugin and turn it off

**Reply by jameserie** [reply]
This solved this issue. https://onebite.dev/solution-for-preflight-cors-issue-in-serverless-function/

I have used vercel.json and added this code on every api I had.

`if(req.method === 'OPTIONS') {
        return res.status(200).json(({
            body: "OK"
        }))
    }`

**Reply by Emmijozzy** [reply]
Pls any solution to the above issue

**Comment by maxschnaider** [comment]
well for me the problem was adblocker

**Comment by stevehusmann** [comment]
Check and see if you’re running into a Captcha. It turned out that was my
issue, and actually had nothing to do with CORS

On Tue, Aug 2, 2022 at 2:34 PM OGUNSUYI JOSEPH OLUWASEUN <
***@***.***> wrote:

> Pls any solution to the above issue
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/vercel/discussions/65#discussioncomment-3310790>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AURF3AP2TXMVCWCAUUYISGDVXFZ6FANCNFSM5RTBCQXA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
Sent from Gmail Mobile

**Reply by Emmijozzy** [reply]
Alright thanks so much, let put a try

**Comment by holylander** [comment]
Besides applying the [CORS config for vercel, described here](https://vercel.com/support/articles/how-to-enable-cors#enabling-cors-using-vercel.json), you will also need your vercel lambda function to return a `200` status code for request using the `OPTIONS` method.

Please also that your **vercel config needs to allow any custom header that you may be using** when querying the endpoint. Lets say your GET request includes a header `X-UserSession`, such header should be allowed by CORS config as:

```
{ "key": "Access-Control-Allow-Headers", "value": "X-UserSession" }
```

**Comment by AJSihota** [comment]
Ive tried the vercel.json approach and allowing the headers in my server code (node.js) but this issue still exists. It's only reproducible in Vercel - when i point my app to my local server everything works...

**Comment by rafakwolf** [comment]
Trying this as well https://vercel.com/guides/how-to-enable-cors
No success

**Comment by AJSihota** [comment]
So frustrating..

**Comment by ugurberkecan** [comment]
tried this https://vercel.com/guides/how-to-enable-cors and  https://onebite.dev/solution-for-preflight-cors-issue-in-serverless-function/

I have used vercel.json and added this code on every api I had.

if(req.method === 'OPTIONS') { return res.status(200).json(({ body: "OK" })) }


No success.

**Comment by likeavenus** [comment]
Not working ALL methonds from https://vercel.com/guides/how-to-enable-cors (???? HOW)

**Comment by amyegan** [comment]
This discussion is a couple of years old at this point, and it's getting a bit difficult to follow with so many comments. 😅  

I'm going to lock this one now, but please don't take that to mean we don't want to help. There are may different reasons why CORS configuration might not be working for a particular app. If the [general guide](https://vercel.com/guides/how-to-enable-cors) is not working for you, please [start a new discussion](https://github.com/vercel/community/discussions/new?category=help) with details and examples from your case so we can help find a solution that will work for your site 🙏
